django/core/serializers/base.py

```diff
@@ -209,6 +209,51 @@ def getvalue(self):
         if callable(getattr(self.stream, "getvalue", None)):
             return self.stream.getvalue()
 
+    def _should_include_pk(self, obj):
+        """
+        Determines whether the Primary Key (PK) should be included in the
+        serialized data for the given object, especially when
+        use_natural_primary_keys=True is set.
+        """
+        if not self.use_natural_primary_keys:
+            return True
+
+        return not self._resolve_natural_key(obj)
+
+    def _resolve_fk_natural_key(self, obj, field):
+        """
+        Return the natural key for a ForeignKey's related object when valid.
+
+        When natural foreign keys are enabled, the helper attempts to use the
+        related object's natural key; otherwise it falls back to the PK.
```

Comment:
I'm not seeing where this fallback behavior is implemented. I'll take a quick stab at simplifying the helpers a little further.